### PR TITLE
just_install_functions

### PR DESCRIPTION
**Pull request by drewgilliam**
New just targets for `just_install_functions` to create isolated sandbox installations of python (installed via conda), pipenv (installed via `30_get-pipenv` an isolated virtualenv), and cmake.  Intended use is to create isolated environments (primarily for windows) that do not conflict with system-level programs.

### CODE REVIEW

File: linux/just_files/just_install_functions.bsh [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+
+  # outputs: pipenv executable & version
+  pipenv_exe="${PIPENV}"
+  pipenv_version="$("${PIPENV}" --version | awk '{print $3}' | sed 's|$\r||g' )"
+  echo "Pipenv ${pipenv_version} installed at \"${pipenv_exe}\""
+
+  # Make sure pipenv meets requirements
+  if ! meet_requirements "${pipenv_version}" "==${pipenv_ver}"; then
+    echo "Pipenv version ${pipenv_version} does not match request ${pipenv_ver}" >&2
+    exit 1
```
Comment (andyneff) [inline linux/just_files/just_install_functions.bsh:409]:
```
JUST_IGNORE_EXIT_CODES=1
return 1
```

File: linux/just_files/just_install_functions.bsh [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+  parse_args pipenv_extra_args \
+      --dir install_dir: \
+      --ver pipenv_ver: \
+      --python PYTHON: \
+      -- ${@+"${@}"}
+
+  # check for pipenv directory
+  if [ -z "${install_dir}" ]; then
+    echo "Pipenv install must specify --dir" >& 2
+    exit 1
```
[the remark this answers is not in the document]
Comment (drewgilliam) [reply linux/just_files/just_install_functions.bsh]: all `exit 1` replaced with the `local JUST_IGNORE_EXIT_CODES=1; return 1` pattern

File: linux/just_files/just_install_functions.bsh [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+  fi
+
+  # executable
+  local cmake_dir="${install_dir}/${cmake_id}"
+  local cmake_exe_footer
+  if [ "${OS-}" = "Windows_NT" ]; then
+    cmake_exe_footer="bin/cmake.exe"
+  else
+    cmake_exe_footer="bin/cmake"
+  fi
```
[the remark this answers is not in the document]
Comment (drewgilliam) [reply linux/just_files/just_install_functions.bsh]: simplified to `/bin/cmake` in all cases, but kept the `cmake_exe_footer` variable as its used in the activation file.

File: linux/just_files/just_install_functions.bsh [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+
+  # install python
+  echo "Installing python..."
+  "${CONDA}" create -y -p "${python_dir}" "python==${python_ver}"
+  local python_exe_footer
+  if [ "${OS-}" = "Windows_NT" ]; then
+    python_exe_footer="python.exe"
+  else
+    python_exe_footer="bin/python"
```
[the remark this answers is not in the document]
Comment (drewgilliam) [reply linux/just_files/just_install_functions.bsh]: Added check for `python3` on linux.

**Comment by drewgilliam** [comment]
@andyneff - I also added tests for the install functions, currently only enabled on appveyor for Windows.  The tests can run on linux machines, but don't play well with alpine (like all the bash dockers) and need at least one file retrieval utility (e.g., wget or curl).  If tests on linux are necessary in the future, maybe can add appropriate utilities to the OS dockers and update the `skip_next_test` check.

**Comment by drewgilliam** [comment]
appveyor seems to be failing on git_mirror tests (should be unrelated to this PR).

File: tests/int/test-just_install_functions.bsh [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+if [ -z ${VSI_COMMON_DIR+set} ]; then
+  VSI_COMMON_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")/.."; pwd)"
+fi
+
+source "${VSI_COMMON_DIR}/tests/testlib.bsh"
+source "${VSI_COMMON_DIR}/linux/just_files/just_install_functions.bsh"
+
+# test cmake install
+[ "${OS-}" = "Windows_NT" ] || skip_next_test
```
Comment (andyneff) [inline tests/int/test-just_install_functions.bsh:13]: Why only in windows?

File: tests/int/test-just_install_functions.bsh [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  # activate version
+  RESULT="$(source "${cmake_activate}"; "${CMAKE_EXE}" --version)"
+  [ "$RESULT" = "${CMAKE_VER_FULL}" ]
+
+)
+end_test
+
+
+# test conda-python-install & pipenv-install
+# test in sequence according to typical usage - install oython then pipenv
```
Comment (andyneff) [inline tests/int/test-just_install_functions.bsh:47]: oython->python

File: tests/int/test-just_install_functions.bsh [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+[ "${OS-}" = "Windows_NT" ] || skip_next_test
```
Comment (andyneff) [inline tests/int/test-just_install_functions.bsh:48]: Why only test in windows?

File: tests/int/test-just_install_functions.bsh (hunk 5 above)
Comment (andyneff) [reply tests/int/test-just_install_functions.bsh:13]: Check for curl/wget instead, like `command -v curl &> /dev/null || command -v wget &> /dev/null || skip_next_test`

File: linux/just_files/just_install_functions.bsh [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+  fi
+
+  # executable
+  local cmake_dir="${install_dir}/${cmake_id}"
+  local cmake_exe_footer="/bin/cmake"
+  local CMAKE="${cmake_dir}/${cmake_exe_footer}"
+
+  # outputs: executable & version
+  cmake_exe="${CMAKE}"
+  cmake_version="$("${CMAKE}" --version | awk 'NR==1{print $3}')"
```
[the remark this answers is not in the document]
Comment (drewgilliam) [reply linux/just_files/just_install_functions.bsh:111]: Just leaving this one as-is

**Pull request merged**